### ISSUE TITLE
Updating (changing) sslcertkey fails

### ISSUE THREAD

**Issue by ajsiersema** [citrix/terraform-provider-netscaler#43]
Following up on [only send nodomaincheck on certificate change #39:](https://github.com/citrix/terraform-provider-netscaler/issues/39) I'm trying to change the certificate by uploading .cert and .key files with the same filename, but different content.
I'm using `local-exec` to actually upload the certificate files.

EXPECTED RESULT:
Changed sslcertkey object

ACTUAL RESULT:
`Error: Error applying plan`

STEPS TO REPRODUCE:

**resources.tf**
```
resource "null_resource" "destroy_certificate" {
  count = "${var.ingress["service_type"] == "SSL" ? 1 : 0}"

  provisioner "local-exec" {
    when = "destroy"
    command = "./cert_files.sh absent"
    working_dir = "${var.cwd}"
    environment = {
      TF_VAR_ns_ingress_name = "${var.ingress["name"]}"
    }
  }
}


resource "null_resource" "upload_certificate" {
  count = "${var.ingress["service_type"] == "SSL" ? 1 : 0}"

  triggers = {
    cert_hash = "${sha1(var.ingress["cert"])}"
  }

  provisioner "local-exec" {
    command = "./cert_files.sh absent"
    working_dir = "${var.cwd}"
    environment = {
      TF_VAR_ns_ingress_name = "${var.ingress["name"]}"
    }
  }
  provisioner "local-exec" {
    command = "./cert_files.sh present"
    working_dir = "${var.cwd}"
    environment = {
      TF_VAR_ns_ingress_name = "${var.ingress["name"]}"
    }
  }
}


resource "netscaler_sslcertkey" "lbsslcertkey" {
  count = "${var.ingress["service_type"] == "SSL" ? 1 : 0}"

  certkey = "TLS_${var.ingress["name"]}_${var.serial}"
  cert = "/nsconfig/ssl/${var.ingress["name"]}.cert"
  key = "/nsconfig/ssl/${var.ingress["name"]}.key"
  expirymonitor = "DISABLED"
  # notificationperiod = 90
  inform = "PEM"
  nodomaincheck = true

  depends_on = ["null_resource.upload_certificate"]
}


resource "netscaler_lbmonitor" "lbmonitor" {
  monitorname = "mon_${var.ingress["name"]}"
  type = "${var.backends_config["service_type"]}"
  interval = 350
  resptimeout = 250
}


resource "netscaler_server" "backends" {
  count = "${length(var.backends)}"

  name = "srv_${format("%s_%s", var.ingress["name"], replace(var.backends[count.index], "/[:.]+/", ""))}"
  ipaddress = "${var.backends[count.index]}"
}


resource "netscaler_servicegroup" "backend" {
  servicegroupname = "sg_${var.ingress["name"]}"
  lbvservers = ["${netscaler_lbvserver.lbvserver.name}"]
  servicetype = "${var.backends_config["service_type"]}"
  servicegroupmembers_by_servername = ["${formatlist("%s:%s", netscaler_server.backends.*.name, var.backends_config["port"])}"]
}


resource "netscaler_lbvserver" "lbvserver" {
  name = "lbvs_${var.ingress["name"]}"
  ipv46 = "${var.ingress["ip"]}"
  port = "${var.ingress["port"]}"
  servicetype = "${var.ingress["service_type"]}"
  lbmethod = "ROUNDROBIN"
  persistencetype = "COOKIEINSERT"
  # ugly workaround to conditionally reference a possible 0 count resource:
  sslcertkey = "${element(concat(netscaler_sslcertkey.lbsslcertkey.*.certkey, list("")), 0)}"
}
```

**variables.tf**

```
variable "ns_user" {
  default = "nsroot"
}

variable "ns_password" {
  default = "nsroot"
}

variable "ns_hostname" {
  default = "ns.domain.tld"
}

variable "cwd" {
  default = "."
}

variable "serial" {
  default = ""
}

variable "ingress" {
  type = "map"
  "description" = "name, vip, port and servicetype for the load balancer"
  default {
    port = 443
    service_type = "SSL"
  }
}

variable "backends_config" {
  type = "map"
  "description" = "port and servicetype for the backend services"
  default {
    clttimeout = 40
    backend_port = 80
    service_type = "HTTP"
  }
}

variable "backends" {
  type = "list"
  description = "List of backend services [ipaddress:port]"
}
```

```
ingress = {
  name = "ademoapp"
  service_type = "SSL"
  ip = "10.42.42.100"
  port = 443

  cert = <<EOF
-----BEGIN CERTIFICATE-----
<snip>
-----END CERTIFICATE-----
EOF

}

backends_config = {
  clttimeout = 40
  port = 80
  service_type = "HTTP"
}

backends = [
  "10.42.42.1",
  "10.42.42.2",
  "10.42.42.3",
]
```

**Create initial setup**

```
export TF_CLI_CONFIG_FILE=terraformrc
export TF_VAR_cwd=/home/user/netscaler/terraform
export TF_VAR_ns_hostname=ns.domain.tld
export TF_VAR_ns_ingress_name=ademoapp
export TF_VAR_ns_password=Pssssh
export TF_VAR_ns_user=my_username
export TF_VAR_serial=9b464bfc0efeb1fd

terraform apply -parallelism=1 -refresh=true -no-color \
  -state app_states/ademoapp.tfstate \
  -var-file=app_vars/ademoapp.tfvars
```

**Resulting state**

terraform show app_states/ademoapp.tfstate

```
netscaler_lbmonitor.lbmonitor:
  id = mon_ademoapp
  action = Not applicable
  alertretries = 0
  application =
  attribute =
  basedn =
  binddn =
  customheaders =
  database =
  destip =
  destport = 0
  dispatcherip = 0.0.0.0
  dispatcherport = 0
  domain =
  downtime = 30
  evalrule =
  failureretries = 0
  filename =
  filter =
  group =
  hostipaddress =
  hostname =
  httprequest = HEAD /
  inbandsecurityid = NO_INBAND_SECURITY
  interval = 350
  iptunnel = NO
  kcdaccount =
  lasversion =
  logonpointname =
  lrtm =
  metric =
  metrictable =
  metricthreshold = 0
  metricweight = 0
  monitorname = mon_ademoapp
  mssqlprotocolversion =
  netprofile =
  oraclesid =
  originhost =
  originrealm =
  password =
  productname =
  query =
  querytype =
  radaccountsession =
  radapn =
  radframedip = 0.0.0.0
  radkey =
  radmsisdn =
  radnasid =
  radnasip = 0.0.0.0
  recv =
  respcode.# = 1
  respcode.0 = 200
  resptimeout = 250
  retries = 3
  reverse = NO
  rtsprequest = HEAD /
  scriptargs =
  scriptname =
  secondarypassword =
  secure = NO
  send =
  servicegroupname =
  servicename =
  sipmethod = OPTIONS
  sipreguri =
  sipuri =
  sitepath =
  snmpcommunity =
  snmpoid =
  snmpthreshold =
  snmpversion = V1
  sqlquery =
  sslprofile =
  state = ENABLED
  storedb = DISABLED
  storefrontacctservice = NO
  storefrontcheckbackendservices = NO
  storename =
  successretries = 1
  tos = NO
  tosid = 0
  transparent = NO
  trofsstring =
  type = HTTP
  units1 = SEC
  units2 = SEC
  units3 = SEC
  units4 = SEC
  username =
  validatecred =
  vendorspecificvendorid = 0
netscaler_lbvserver.lbvserver:
  id = lbvs_ademoapp
  appflowlog = ENABLED
  authentication = OFF
  authenticationhost =
  authn401 = OFF
  authnprofile =
  authnvsname =
  backuplbmethod =
  backuppersistencetimeout = 2
  backupvserver =
  bypassaaaa = NO
  cacheable = NO
  comment =
  connfailover = DISABLED
  cookiename =
  dbprofilename =
  dbslb =
  disableprimaryondown = DISABLED
  dns64 = DISABLED
  dnsprofilename =
  downstateflush = ENABLED
  hashlength = 0
  httpprofilename =
  httpsredirecturl =
  icmpvsrresponse = PASSIVE
  insertvserveripport = OFF
  ipmask = *
  ippattern = 0.0.0.0
  ipset =
  ipv46 = 10.42.42.100
  l2conn = OFF
  lbmethod = ROUNDROBIN
  lbprofilename =
  listenpolicy = NONE
  listenpriority = 0
  m = IP
  macmoderetainvlan = DISABLED
  mssqlserverversion =
  mysqlcharacterset = 0
  mysqlprotocolversion = 0
  mysqlservercapabilities = 0
  mysqlserverversion =
  name = lbvs_ademoapp
  netmask =
  netprofile =
  newname =
  newservicerequest = 0
  newservicerequestincrementinterval = 0
  newservicerequestunit = PER_SECOND
  oracleserverversion =
  persistencebackup = NONE
  persistencetype = COOKIEINSERT
  persistmask = 255.255.255.255
  port = 443
  pq =
  processlocal = DISABLED
  push = DISABLED
  pushlabel = none
  pushmulticlients = NO
  pushvserver =
  recursionavailable =
  redirectfromport = 0
  redirectportrewrite = DISABLED
  redirurl =
  redirurlflags = false
  resrule =
  retainconnectionsoncluster = NO
  rhistate = PASSIVE
  rtspnat = OFF
  rule =
  sc =
  servicename =
  servicetype = SSL
  sessionless = DISABLED
  skippersistency = None
  sobackupaction =
  somethod = NONE
  sopersistence = DISABLED
  sothreshold = 0
  sslcertkey = TLS_ademoapp_9b464bfc0efeb1fd
  sslprofile =
  state =
  tcpprofilename =
  timeout = 2
  tosid = 0
  trofspersistence = ENABLED
  v6netmasklen = 0
  vipheader =
  weight = 0
netscaler_server.backends.0:
  id = srv_ademoapp_1042421
  comment =
  delay = 0
  domain =
  domainresolvenow = false
  domainresolveretry = 0
  graceful =
  internal = false
  ipaddress = 10.42.42.1
  ipv6address = NO
  name = srv_ademoapp_1042421
  newname =
  state = ENABLED
  translationip = 0.0.0.0
  translationmask = 0.0.0.0
netscaler_server.backends.1:
  id = srv_ademoapp_1042422
  comment =
  delay = 0
  domain =
  domainresolvenow = false
  domainresolveretry = 0
  graceful =
  internal = false
  ipaddress = 10.42.42.2
  ipv6address = NO
  name = srv_ademoapp_1042422
  newname =
  state = ENABLED
  translationip = 0.0.0.0
  translationmask = 0.0.0.0
netscaler_server.backends.2:
  id = srv_ademoapp_1042423
  comment =
  delay = 0
  domain =
  domainresolvenow = false
  domainresolveretry = 0
  graceful =
  internal = false
  ipaddress = 10.42.42.3
  ipv6address = NO
  name = srv_ademoapp_1042423
  newname =
  state = ENABLED
  translationip = 0.0.0.0
  translationmask = 0.0.0.0
netscaler_servicegroup.backend:
  id = sg_ademoapp
  appflowlog = ENABLED
  autodisabledelay = 0
  autodisablegraceful =
  autoscale = DISABLED
  cacheable = NO
  cachetype = SERVER
  cip = ENABLED
  cipheader = X-Forwarded-For
  cka = NO
  clttimeout = 180
  cmp = NO
  comment =
  customserverid =
  delay = 0
  downstateflush = ENABLED
  dupweight = 0
  graceful = NO
  healthmonitor = YES
  httpprofilename =
  includemembers = false
  lbmonitor =
  lbvservers.# = 1
  lbvservers.4129248370 = lbvs_ademoapp
  memberport = 0
  monconnectionclose = NONE
  monitornamesvc =
  netprofile =
  newname =
  pathmonitor = NO
  pathmonitorindv = NO
  port = 0
  riseapbrstatsmsgcode = 0
  rtspsessionidremap = OFF
  sc =
  serverid = 0
  servername =
  servicegroupmembers.# = 0
  servicegroupmembers_by_servername.# = 3
  servicegroupmembers_by_servername.1839026872 = srv_ademoapp_1042423:80:1
  servicegroupmembers_by_servername.2797699357 = srv_ademoapp_1042422:80:1
  servicegroupmembers_by_servername.542503859 = srv_ademoapp_1042421:80:1
  servicegroupname = sg_ademoapp
  servicetype = HTTP
  sp = OFF
  state = ENABLED
  svrtimeout = 360
  tcpb = NO
  tcpprofilename =
  useproxyport = YES
  usip = NO
  weight = 0
netscaler_sslcertkey.lbsslcertkey:
  id = TLS_ademoapp_9b464bfc0efeb1fd
  bundle =
  cert = /nsconfig/ssl/ademoapp.cert
  certkey = TLS_ademoapp_9b464bfc0efeb1fd
  expirymonitor = DISABLED
  fipskey =
  hsmkey =
  inform = PEM
  key = /nsconfig/ssl/ademoapp.key
  linkcertkeyname =
  nodomaincheck = false
  notificationperiod = 0
  ocspstaplingcache = false
  passplain =
  password = false
null_resource.destroy_certificate:
  id = 2452242766656250296
null_resource.upload_certificate:
  id = 4257515104869952010
  triggers.% = 1
  triggers.cert_hash = 66b0e9969c211e5a824b504d092802e57d9f3367
```

**Changes**
- contents of certificate files : app_vars/ademoapp.cert, app_vars/ademoapp.key
- cert var variable
- name of lbsslcertkey to trigger change ("taint"), see TF_VAR_serial

_app_vars/ademoapp.tfvars_
```

ingress = {
  ....
  cert = <<EOF
-----BEGIN CERTIFICATE-----
MIIDmTCCAoGgAwIBAgIJAJtGS/wO/rH9MA0GCSqGSIb3DQEBCwUAMGsxCzAJBgNV
<snip>
-----END CERTIFICATE-----
...
}
```

_Execute terraform again_
```

export TF_CLI_CONFIG_FILE=terraformrc
export TF_VAR_cwd=/home/user/netscaler/terraform
export TF_VAR_ns_hostname=ns.domain.tld
export TF_VAR_ns_ingress_name=ademoapp
export TF_VAR_ns_password=Pssssh
export TF_VAR_ns_user=my_username
export TF_VAR_serial=123456789abcdef
terraform apply -parallelism=1 -refresh=true -no-color \
  -state app_states/ademoapp.tfstate \
  -var-file=app_vars/ademoapp.tfvars

```

```
null_resource.upload_certificate: Refreshing state... (ID: 4257515104869952010)
null_resource.destroy_certificate: Refreshing state... (ID: 2452242766656250296)
netscaler_server.backends[2]: Refreshing state... (ID: srv_ademoapp_1042423)
netscaler_server.backends[1]: Refreshing state... (ID: srv_ademoapp_1042422)
netscaler_server.backends[0]: Refreshing state... (ID: srv_ademoapp_1042421)
netscaler_lbmonitor.lbmonitor: Refreshing state... (ID: mon_ademoapp)
netscaler_sslcertkey.lbsslcertkey: Refreshing state... (ID: TLS_ademoapp_9b464bfc0efeb1fd)
netscaler_lbvserver.lbvserver: Refreshing state... (ID: lbvs_ademoapp)
netscaler_servicegroup.backend: Refreshing state... (ID: sg_ademoapp)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place
-/+ destroy and then create replacement

Terraform will perform the following actions:

~ netscaler_lbvserver.lbvserver
      sslcertkey:                                   "TLS_ademoapp_9b464bfc0efeb1fd" => "TLS_ademoapp_123456789abcdef"

  ~ netscaler_servicegroup.backend
      servicegroupmembers_by_servername.1839026872: "srv_ademoapp_1042423:80:1" => ""
      servicegroupmembers_by_servername.213590207:  "" => "srv_ademoapp_1042423:80"
      servicegroupmembers_by_servername.2796742708: "" => "srv_ademoapp_1042421:80"
      servicegroupmembers_by_servername.2797699357: "srv_ademoapp_1042422:80:1" => ""
      servicegroupmembers_by_servername.3020376026: "" => "srv_ademoapp_1042422:80"
      servicegroupmembers_by_servername.542503859:  "srv_ademoapp_1042421:80:1" => ""

  ~ netscaler_sslcertkey.lbsslcertkey
      certkey:                                      "TLS_ademoapp_9b464bfc0efeb1fd" => "TLS_ademoapp_123456789abcdef"
      nodomaincheck:                                "false" => "true"

-/+ null_resource.upload_certificate (new resource required)
      id:                                           "4257515104869952010" => <computed> (forces new resource)
      triggers.%:                                   "1" => "1"
      triggers.cert_hash:                           "66b0e9969c211e5a824b504d092802e57d9f3367" => "80c5376a34359caf831c1bef1724bb296764d1cd" (forces new resource)
Plan: 1 to add, 3 to change, 1 to destroy.



null_resource.upload_certificate: Destroying... (ID: 4257515104869952010)
null_resource.upload_certificate: Destruction complete after 0s
null_resource.upload_certificate: Creating...
  triggers.%:         "" => "1"
  triggers.cert_hash: "" => "80c5376a34359caf831c1bef1724bb296764d1cd"
null_resource.upload_certificate: Provisioning with 'local-exec'...
null_resource.upload_certificate (local-exec): Executing: ["/bin/sh" "-c" "./cert_files.sh absent"]
null_resource.upload_certificate: Provisioning with 'local-exec'...
null_resource.upload_certificate (local-exec): Executing: ["/bin/sh" "-c" "./cert_files.sh present"]
null_resource.upload_certificate: Creation complete after 12s (ID: 8026929991478337533)
netscaler_sslcertkey.lbsslcertkey: Modifying... (ID: TLS_ademoapp_9b464bfc0efeb1fd)
  certkey:       "TLS_ademoapp_9b464bfc0efeb1fd" => "TLS_ademoapp_123456789abcdef"
  nodomaincheck: "false" => "true"
netscaler_sslcertkey.lbsslcertkey: Modifications complete after 0s (ID: TLS_ademoapp_9b464bfc0efeb1fd)

Error: Error applying plan:

1 error(s) occurred:

* netscaler_lbvserver.lbvserver: netscaler_lbvserver.lbvserver: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.11.13
    Resource ID: netscaler_lbvserver.lbvserver
    Mismatch reason: attribute mismatch: sslcertkey
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"sslcertkey":*terraform.ResourceAttrDiff{Old:"TLS_ademoapp_9b464bfc0efeb1fd", New:"TLS_ademoapp_123456789abcdef", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.


```

**Comment by giorgos-nikolopoulos** [comment]
I verified the bug you reported.

The root cause is the fact that the `netscaler_sslcertkey` resource in the planning phase appears to be updated with only the name changing.
The actual code in the module does nothing when only this attribute is changed, so in the application phase the update was a noop.
Subsequently the key referenced by the `netscaler_lbvserver` resource was non existent so the diff was different from the plan and that caused the terraform framework to fail with the error message you reported.

The fix is implemented in #44 where we mark the `certkey` attribute with the `ForceNew` flag to force a deletion and recreation of the resource, which is the correct behavior.

The fix is also present in the latest release [0.11.13](https://github.com/citrix/terraform-provider-netscaler/releases/tag/v0.11.13)

For your terraform plan to work you will also need to add the lifecycle option `create_before_destroy` in the `netscaler_sslcertkey` resource.

This is  needed to avoid a cyclical dependency between the sslcertkey and the lbvserver resource.
Without this option terraform will first try to delete the old sslcertkey which will fail since it is referenced by the existing lbvserver.
Adding this lifecycle optoin avoids this problem since it performs all the updates before deleting the old certificate.

I have a sample resource script below that I used to verify the correctness of this operation.

```
resource "netscaler_sslcertkey" "lbsslcertkey" {
  certkey = "certificate_mock_1"
  cert = "/nsconfig/ssl/certificate_mock.pem"
  key = "/nsconfig/ssl/key_mock.pem"
  expirymonitor = "DISABLED"
  nodomaincheck = true
  inform = "PEM"
  lifecycle {
    create_before_destroy = true
  }

}


resource "netscaler_lbvserver" "lbvserver" {
  name = "ssltest_lbvserver"
  ipv46 = "10.78.1.34"
  port = "443"
  servicetype = "SSL"
  lbmethod = "ROUNDROBIN"
  persistencetype = "COOKIEINSERT"
  sslcertkey = "${netscaler_sslcertkey.lbsslcertkey.certkey}"
}

```

@ajsiersema let us know if this works out for you.

**Comment by ajsiersema** [comment]
@giorgos-nikolopoulos ,

Thanks for the speedy reply and the explanation.
I've changed the resource script to include the lifecycle option, in fact, that's the only thing I've changed in my resources.tf.  However, it still fails, see below. I can't spot the difference with your test setup, or am I missing something else ?

```
resource "netscaler_sslcertkey" "lbsslcertkey" {
  count = "${var.ingress["service_type"] == "SSL" ? 1 : 0}"
  certkey = "TLS_${var.ingress["name"]}_${var.serial}"
  cert = "/nsconfig/ssl/${var.ingress["name"]}.cert"
  key = "/nsconfig/ssl/${var.ingress["name"]}.key"
  expirymonitor = "DISABLED"
  inform = "PEM"
  nodomaincheck = true

  lifecycle {
    create_before_destroy = true
  }

  depends_on = ["null_resource.upload_certificate"]
}

resource "netscaler_lbvserver" "lbvserver" {
  name = "lbvs_${var.ingress["name"]}"
  ipv46 = "${var.ingress["ip"]}"
  port = "${var.ingress["port"]}"
  servicetype = "${var.ingress["service_type"]}"
  lbmethod = "ROUNDROBIN"
  persistencetype = "COOKIEINSERT"
  sslcertkey = "${element(concat(netscaler_sslcertkey.lbsslcertkey.*.certkey, list("")), 0)}"
}

```

It still fails:

```
1 error(s) occurred:

* netscaler_lbvserver.lbvserver: netscaler_lbvserver.lbvserver: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.11.13
    Resource ID: netscaler_lbvserver.lbvserver
    Mismatch reason: attribute mismatch: sslcertkey
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{\"sslcertkey\":*terraform.ResourceAttrDiff{Old:\"TLS_ademoapp_ae137539706f54de\", New:\"TLS_ademoapp_9b464bfc0efeb1fd\", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}

```

**Comment by giorgos-nikolopoulos** [comment]
I changed my terraform version to match yours 0.11.13 and I added the dependency to the null_resource in case this was creating some different kind of flow.

It still works correctly for me.

The only thing I can think of is that the netscaler provider in the .terraform folder inside the configuration scripts folder is not updated to the 0.11.13 version.
Check the `lock.json` inside the `.terraform` folder to have a matching sha256sum with the binary from the release for the netscaler provider.

If this still fails I would need the whole output from the `terraform apply` to see if I can find any other possible sources of error.
